### DISCUSSION TITLE
Speeding Python RDKit function calls on many molecules--release GIL, vectorize?

### DISCUSSION THREAD

**Discussion by bertiewooster** [rdkit/rdkit#7611] [Development]
When using the Python API, often one wants to perform the same RDKit operation on many molecules (or SMILES strings, etc.). For example, if one wants to convert the ~100 million SMILES in [PubChem](https://pubchem.ncbi.nlm.nih.gov/) into RDKit molecules, one would call `[Chem.MolFromSmiles(sml) for sml in pubchem_smls]` where `pubchem_smiles` is an iterable of those ~100M SMILES strings. For 1 million SMILES this takes about 6 seconds on my computer, so the time becomes significant for 100 million (expect 600 seconds = 10 minutes).

As I understand it, while each C++ function call is fast, the Python code runs slowly because each of the ~100M calls requires the global interpreter lock (GIL).

Is there already a faster way (comparable to C++ calls) than the Python code above? If not, is it possible to release the GIL for this and similar RDKit functions, as @bp-kelley suggested below? Or would it be feasible to add vector functions so one Python call could pass many e.g. SMILES to a C++ function, for example a hypothetical `Chem.MolFromSmilesMulti(pubchem_smiles)` that would return an iterable of molecules?

---

[From](https://github.com/rdkit/rdkit/discussions/7134#discussioncomment-8361774) @bp-kelley:

> We also can do another sweep of functions that should release the GIL.  I think standardization functions tautomer enumeration are good candidates for this.
> 
> We do it in a few places in the standardize library, but not everywhere:
> 
> https://github.com/rdkit/rdkit/blob/8dae48b7a17fd984c69d04549e6d9b53690f5c52/Code/GraphMol/MolStandardize/Wrap/rdMolStandardize.cpp#L190

**Comment by kienerj** [comment]
Is it really python that is slow or do you really now in plain C++ it would be a lot faster? As far as I know creating the molecules is quiet "heavy" operation. from your number if I don't make any mistake it's 6µs per molecule. 
(As reference reading molecules into rdkit using KNIME = java-based, no gil, easily multi-threaded) is also kind of slow. )

The issue I think is simply a costly operation done a lot of times = slow?

Solution would be multi-processing, using multiple processes for the SMILES to RDKit parsing. Like [this](https://github.com/rdkit/rdkit/issues/2529) but adapted to your use-case or even one-step further "stream processing" in case you have memory limits:
[Python patterns for processing large SDF files ](https://baoilleach.blogspot.com/2020/05/python-patterns-for-processing-large.html )
But adjusted for SMILES.

**Comment by bertiewooster** [comment]
Good ideas to try, thanks.

**Reply by kienerj** [reply]
You might also get some inspiration from [rdkit_parallel_tools](https://github.com/kienerj/rdkit_parallel_tools).
